src/sqlfluff/core/templaters/slicers/tracer.py

```diff
@@ -192,7 +192,17 @@ def move_to_slice(
             # Save all the candidates for each step so we can return them later.
             step_candidates[self.program_counter] = candidates
             # Step forward to the best step found.
-            self.program_counter = candidates[0]
+
+            # https://github.com/sqlfluff/sqlfluff/issues/6121
+            # If the target is inside a for loop, we need to move to the
+            # first element inside the loop.
+            if (current_raw_slice.tag == "endfor"
+                # Elements of inside_set_macro_or_call have empty next_slice_indices
+                and self.raw_slice_info[current_raw_slice].next_slice_indices
+                and self.raw_slice_info[current_raw_slice].next_slice_indices[0] <= target_slice_idx < self.program_counter):
+                self.program_counter = self.raw_slice_info[current_raw_slice].next_slice_indices[0]
```

Comment:
There's a lot of a reuse of `self.raw_slice_info[current_raw_slice].next_slice_indices`, and I think it's much easier to understand if that's a local variable. The only but that isn't totally clear is the inequality comparison, could you add a note there on how people should understand the logic there?

```suggestion
            next_slice_indices = self.raw_slice_info[current_raw_slice].next_slice_indices
            if (
                current_raw_slice.tag == "endfor"
                # Elements of inside_set_macro_or_call have empty next_slice_indices
                and next_slice_indices
                # TODO: Add a comment to explain this logic
                and next_slice_indices[0] <= target_slice_idx < self.program_counter
            ):
                self.program_counter = next_slice_indices[0]
```

Reply:
Got it, should be good now, thanks.